firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +60,142 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  static class CallbackList {
+    void run() {
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void add(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private final List<Runnable> callbacks = new ArrayList<>();
+  }
+
+  public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    CallbackList callbacks = new CallbackList();
+  }
+
+  @SuppressWarnings("deprecation")
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    CallbackList callbacks = new CallbackList();
+  }
+
+  @Nullable
+  private static <T> T castFragment(Class<T> fragmentClass, @Nullable Object fragment, String tag) {
+    try {
+      return fragmentClass.cast(fragment);
+    } catch (ClassCastException e) {
+      throw new IllegalStateException(
+          "Fragment with tag '"
+              + tag
+              + "' is a "
+              + fragment.getClass().getName()
+              + " but should be a "
+              + fragmentClass.getName());
+    }
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private static WeakHashMap<Activity, WeakReference<StopListenerSupportFragment>>
```

Comment:
WeakHashMap isn't quite right, or if it is we need to be significantly more careful about clearing listeners once done.

The problem is that the callbacks the listener takes are closures that can capture references to the Activity. WeakHashMap holds values by strong reference which means that our map will never be collected and we'll prevent garbage collection of user activities.

This could possibly be mitigated by clearing the callback list after running them?

Reply:
This (and other related comments) are no longer relevant. Unless we decide not to sync on the ui thread.